Enhancement: Node.js Local Document Generator

3DEsprit opened the pull request:
Update One adding custom Node.js documentation building, to start the upgrade process per Issue #503 .

The new API module is setup to create new documents once learnyounode is started, generating to node_apidoc and the version reflecting the version of Node.js used to execute. To keep with the consistency of offering an offline solution, all requests are made locally so there is no need for connectivity after the deps have been installed.

_Note: Once the documents are created any future requests will be ignored as long as the folder exists, making sure not to add overhead on every single run. If a user accidentally removes the doc folder, it will recreate that folder on the next run._ 

To remove the need to overwrite files, reduce size, and redundancy the old 0.10 apidoc folder has been removed.

3DEsprit commented:
The CI build check is choking on the docbuild file which cannot get past the test. I am not sure what to do to get around the errors on the require calls, but everything works as intended.

AnshulMalik commented on lib/docbuild.js:
```diff
@@ -0,0 +1,7 @@
+createDocs = require('node-offline-api').createDocs
+buildOptions = require('node-offline-api').buildOptions
```
can we use something like `var buildOptions = require('')` or `const` may be.

That would help the test to pass.

3DEsprit replied:
Thanks @AnshulMalik !

Wow, what in the world was I doing when I submitted that? I think I spent a little too much time with Ruby that day. :) 

I has been done! Using var for now, until the lowest version of Node is a specific, then I can cross-check the available ES6 sugar that can be safely used.

3DEsprit commented:
Looks like one more issue where it is checking the local file build, and adding the default folder path. Fix on the way up.

AnshulMalik commented:
Looks like the travis is still unhappy :P

3DEsprit commented:
Ok, it looks like it is now failing on the official Node.js document generator API that is part of the offline document API, otherwise it seems to be working as intended. I could go through those files, and change them all to standard if needed, and it should remove any of the CI errors.

3DEsprit commented:
Ok, I see now. It is catching the fs.stat error which isn't an actual error. Let me quiet that down, and see if it will ignore the standard on the official after that. Fix coming up.

3DEsprit commented:
Well, the fatal errors are gone now. The only issue with the tests is where the tests see some of the verbosity of the API when testing the exercises. 

The reason for this is due to running that offline document API upon startup. It is run quietly per the _buildOptions.quiet_ setting so no one actually sees this when running LYN, and it only runs when the document directory does not exist. For some reason the exercise tests still see these in the test though. 

If this is something that just has to pass, I can make another release without these messages all together.

3DEsprit commented:
Ok, refactored, and re-released module to allow 100% passing locally.

3DEsprit commented:
It looks like there is a possible issue when validating files. I'm going to take the weekend to look for a way to limit the possible executions, instead of running on every instance.

3DEsprit commented:
Since I rebased with the latest master, it seems that the 5.x build is now erroring, but the 6.x and 7.x tests are passing. Unfortunately the 5.0 build is giving an npm error that if cannot find the function `Buffer.alloc`, which requires Node.js `5.10.0`+ and the CI is using `5.0.0`:

<img width="952" alt="screen shot 2017-12-01 at 19 14 02" src="https://user-images.githubusercontent.com/3156278/33509974-33408692-d6cc-11e7-95fa-7bda3e80ace8.png">

<img width="909" alt="screen shot 2017-12-01 at 19 16 08" src="https://user-images.githubusercontent.com/3156278/33509975-3613ccee-d6cc-11e7-84c0-70dd1e80aed4.png">

Looking through my latest source, I am not finding this being called. It looks like this is an external issue, but not sure where when no node modules are being installed due to the error.

Please let me know if you need anything from me, as I would love to see this move forward so I can continue to help with new updates to help make this a more recent-friendly workshop. 😃

martinheidegger commented on package.json:
```diff
@@ -80,7 +81,7 @@
   },
   "scripts": {
-    "lint": "./node_modules/.bin/standard",
+    "lint": "./node_modules/.bin/standard --fix",
```
This should not be part of THIS pr! (Note: during development you can run `npm run lint -- --fix`

martinheidegger commented on lib/docbuild.js:
```diff
@@ -0,0 +1,10 @@
+buildOptions.buildName = 'node_apidoc'
+
+fs.stat('../node-apidoc', function (err) {
+  if (err) { createDocs() }
```
There can theoretically be various errors here. I think the [`.code === "ENOENT"`](https://nodejs.org/api/errors.html#errors_common_system_errors) is the error we are looking for.

martinheidegger commented on lib/docbuild.js:
```diff
@@ -0,0 +1,10 @@
+buildOptions.buildDir = process.cwd()
+buildOptions.buildName = 'node_apidoc'
+
+fs.stat('../node-apidoc', function (err) {
```
I try to avoid modules without an `export`, can this module also wrap the operation in a module?

martinheidegger commented on bin/learnyounode:
```diff
@@ -1,3 +1,7 @@
\ No newline at end of file
+require('../learnyounode').execute(process.argv.slice(2))
+
+if(process.argv.slice(2) == 0){
```
_(This is not standard syntax)_

Workshopper has a [`commands` option](https://github.com/workshopper/workshopper-adventure/blob/7c0493edc6b4374be60898c08fa23130e65aa7d7/index.js#L102) that allows to register custom commands just for this purpose. Of course there should be a [`postinstall`](https://docs.npmjs.com/misc/scripts) script that launches this command.

3DEsprit replied:
Huge thanks Martin. I'll get these knocked out this week.

3DEsprit commented:
I went ahead and cleaned up the module, and added the postinstall script. Let me know if there is anything else that you feel would work better, or be more ideal in other situations.
